Caching & Playback Timing Improvements

jonoomph opened the pull request:
This PR is a huge refactor related to video and audio **playback** in OpenShot. Details below.

**Improvements and Bug Fixes in this PR:**
 - Replaced very bad video timing code with a much **improved video timing code**. We now track the "time" playback starts, and we can accurately predict the "End time" of any frame # (`start_time + (frame_counter_since_start * frame_duration)`. So, instead of the video thread keeping a "relative" sleep per loop with no way of accurately synchronizing with the audio thread, we are always taking the current time minus the calculated end time, for a 100% perfect video timing. This is the solution I have been searching for. Seems obvious now, lol.
 - **Refactored video cache**, which now works much more reliably than before. It now caches in both directions (forward in time, and backwards), filling in missing frames when detected. It now only uses 1/2 the cache size, so 50% behind the playhead & 50% in front of the playhead. This allows users to easily scrub forward and backwards in the cache, without needing to re-generate frames. 
 - **Fixed a bad audio buffer** design, where we had 2 buffers for no reason. This fixes a bug where audio would stay in one of our audio buffers, and not empty out when seeking to a new position (i.e. playing a tiny bit of sound from the last position).
 - **Lower audio latency** (sending 500 samples per callback now, usually around 88 - 96 callbacks per second). No more pre-buffering in JUCE.
 - **Pre-roll support** before playback. Instead of instantly starting playback, we always wait for our pre-roll to cache a hand-full of frames before playback begins.
 - Improved fast forward / rewind to use correct timing
 - Utilizing **higher resolution timers** (no more milliseconds... which caused issues on certain framerates that needed 1/2 milliseconds.
 - **Fixed big memory leak** in our use of Vectors (in FrameMapper, Cache, Keyframe classes). This was most noticeable on very large Clips (i.e. 60+ minutes duration), plus usage of hundreds of these clips (and hundreds of FrameMappers).... thus creating many, many GB of Vector data, and never really fully releasing the memory until we quit OpenShot.
 - **Improved sample rate detection** (especially on Windows). We output a warning if we see a mismatching sample rate. This is a bit misleading, as on Linux and Mac this mismatch does not really cause any issues (since the audio drivers resample all sounds anyway... if needed). But on Windows, some audio drivers simply don't resample correctly, so we output this message to help troubleshoot if your audio is playing back faster/slower pitch.
 - **Fixed a crash on shutdown** of audio thread

~~This PR also includes some new, experimental timing code to improve our video playback. Historically, our audio thread does not report an accurate position (i.e. in frame position). It is sometimes a few frames ahead, and sometimes not. It is variable, and depends on how many samples are requested from JUCE (and ultimately the audio drivers). Because of this "inaccuracy", our timing logic can not react immediately to audio/video sync... as it might be incorrect. Thus, we are now averaging the sync over many frames, to calculate if we are far enough away to trigger an adjustment. The adjustment now takes place in tiny increments, and over many frames to bring the sync back near zero frames difference.~~

~~This adjustment can cause a subtle oscillation over time, as the adjustment drifts down and up, always chasing zero. But due to the small size of adjustment, it should be undetectable. Also, once we reach zero drift, we disable the adjustment code until it is triggered again... hopefully ignoring the random jitter of the audio position inaccuracies. However, if over time, the average diff keeps increasing/decreasing, it will trigger another adjustment, and so on.~~

~~I have tried dozens of different approaches for correct audio/video sync during playback, and graphed them all for analysis. This approach, by far, has the best results so far... with minimal audio/video desync, and minimal oscillation. So I think it's time for more eyeballs to experiment with playback using this code... for a real-world test.~~

Commit:
- Searched for word 'millisecond' replaced with 'microsecond'
- Simplify timing correction with the cube of the difference between audio and video times.
    - If the correction is longer than it took to generate a frame. skip frames until under a frame behind.

Commit:
skip frames when over a frame_duration behind

Commit:
Better comment, tune the timing

- Spent time tuning the function so it keeps the error (video/audio difference) low
- Broke up the code, and added comments to improve readability.

Commit:
Re-included debug statement

Commit:
Simplify math, add max sleep time.

- Simplified to one formula.
- Removed additional case for skipping frames.

jonoomph committed:
Refactoring the VideoCacheThread to check every frame before requesting it. Adding a new method: Contains() to our cache objects, to facilitate this. Removing cache clearning experimental code from Timeline (causing playback issues). Refactoring PrivatePlayer playback timing code, to calculate an average # of frame difference between audio and video threads, and slowly adjust back towards zero when needed.

Commit:
add logging file

JacksonRG commented:
My bad! made a change and meant to push it to my personal repo to access it tomorrow. I'll undo that commit

jonoomph committed:
Using more aggressive adjustment increment, and adding DEBUG output

jonoomph committed:
Merge remote-tracking branch 'origin/caching-timing-improvements' into caching-timing-improvements

# Conflicts:
#	src/Qt/PlayerPrivate.cpp

jonoomph committed:
Refactoring video playback timing to use absolute timestamps and predicted frame times, instead of relative sleep. Adjusting audio buffer smaller (less latency).

jonoomph committed:
Adding ability to open *.osp files in QtPlayer app.

jonoomph committed:
Adding pre-roll to VideoCacheThread, and making video & audio threads wait for the isReady() method before playing. Also fixed an audio bug where our internal buffer was not being cleared on seek. Removed some unused caching from Clip, and did some minor refactor on FrameMapper/Clip cache clearing.

jonoomph committed:
Big refactor of AudioReaderSource, to be simpler, and only fill the audio samples requested (removed double/internal buffer complexity). Also, we now initialize the audio device manager at a specific sampleRate and # of channels correctly. Removed 'latency' adjustment in PrivatePlayer (no longer needed with no read-ahead buffer). Increased the min_frames_required on the video cache to 24.

jonoomph committed:
Moving from high resolution timer to system_clock, trying to avoid a Mac build error

jonoomph committed:
Improved pausing code on video playback, to allow for scrubbing during paused state.

jonoomph committed:
Break out of cache loop if playback speed changes during caching. Move getFrame() below pause code, to prevent accidental incrementing video_position.

jonoomph committed:
Reducing pre-roll to 12 frames

jonoomph committed:
Fixing memory issues with FrameMapper, Cache, and Keyframe objects (which use vectors that can hold tons of objects, and were not fully releasing the memory on the heap until the program terminates)

jonoomph committed:
Removing audio seek from PlayerPrivate.cpp pause code, since this causes crashes when terminating openshot-qt.

jonoomph committed:
Large refactor to caching and playback timing:
 - Support different speeds (-1X, 2X, 4X, etc...)
 - Clamp getFrame between 1 and timeline length
 - Support rewind using new timing code
 - Caching in both directions (based on previous speed)
 - Removing mutex from Timeline::GetFrame (cached path)
 - Caching supports actual speed now
 - Bust cache and trigger pre-roll if we encounter too many uncached frames. This will re-sync the video + audio threads.
 - Sleep less in cache loop / Adjusting min frames to 24 (trial and error). Too few and backwards playback suffers. Too many and all playback waits.

jonoomph committed:
DEBUG output of default output sample rate

jonoomph committed:
Don't use any input channels when initializing audio device

jonoomph committed:
Detecting default sample rate of default audio device (Windows seems to play audio out of sync on certain audio drivers, if there is a mismatch between OpenShot and the default system audio sample rate). This info will help debug issues. Adding back in a mutex on Timeline::GetFrame causing crashes.

JacksonRG commented on src/Qt/PlayerPrivate.cpp:
```diff
@@ -55,94 +57,75 @@ namespace openshot
+            // Also pause at end of timeline, and pause if 'playing' and
+            // the pre-roll is not yet ready
+            if ((speed == 0 && video_position == last_video_position) ||
+                (speed != 0 && last_speed != speed) ||
```
Noticed we dropped the `(video_position > reader->info.video_length)` check here. Do we no longer need that?

jonoomph committed:
Improved scrubbing to not wait on pre-roll

jonoomph committed:
Prevent crashes:
- Tweaking GetSmallestFrame to not use an additional mutex.
- Replace 2 mutex on Timeline::GetFrame with 1 mutex.
- Remove mutex from Timeline::ClearAllCache.
- Slow down loop on video cache thread.
- Protect audioThread::Seek() if missing a source.
- Rearrange StopPlayback to always stop the videoCache thread first.

jonoomph committed:
Merge branch 'develop' into caching-timing-improvements

jonoomph committed:
Removing test.sh file

jonoomph committed:
Remove debugging code

jonoomph commented:
This branch has finally passed our initial user testing (internally), and is ready for wider testing on the daily build.
